### PR TITLE
:book: Update Branch-Protection admin and non-admin requirements

### PR DESCRIPTION
**Pull request by gabibguti**
#### What kind of change does this PR introduce?

This PR introduces a docs update for Branch-Protection check. It updates:
- which branch protection settings are required using non-admin and admin tokens
- which branch protection setting is evaluated in which Tier
- how to setup a perfect branch protection score for GitHub repos

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

The docs say:
- admin token is required to verify some settings which are available using non-admin token
- the settings evaluated are distributed in a way for the Tier's when it's distributed in a different way

#### What is the new behavior (if this is a feature change)?**

Fixed the settings required for admin and non-admin tokens and the settings distribution by Tier. Also, added a FAQ question to help users to setup a perfect branch protection score for GitHub repos.

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

NONE

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
Update docs on Branch-Protection to fix the settings required when using admin and non-admin tokens and the settings distribution by Tier
```

### CODE REVIEW

File: docs/faq.md
Comment (diogoteles08) [inline docs/faq.md]: Ideally we would like to have alternative texts on those images, following the syntax `alt text`.

However this is a difficult image to describe. In the blogposts, I've ben describing following this template, feel free to reuse:

[GitHub page where we define which protection rules will be applied to the selected branch, with the rules as checkboxes. Read below all the options ordered as they are displayed, followed the info if they're currently flagged or not: 
In the section named as "Protect matching branches", we have "Require a pull request before merging" flagged; "Require approvals" unflagged; "Dismiss stale pull request approvals when new commits are pushed" unflagged; "Require review from Code Owners" flagged; "Require approval of the most recent push" unflagged; "Require status checks to pass before merging" unflagged; "Require conversation resolution before merging" unflagged; "Require signed commits" unflagged; "Require linear history" unflagged; "Require deployments to succeed before merging" unflagged; "Lock branch" unflagged;  "Do not allow bypassing the above settings" flagged. In the section named as "Rules applied to everyone including administrators", we have "Allow force pushes" unflagged; "Allow deletions" unflagged]

File: docs/checks.md
Comment (diogoteles08) [inline docs/checks.md:117]: For code owners Scorecard does not require admin tokens, as far as I have tested.

Comment (diogoteles08) [inline docs/checks.md:117]: For code owners Scorecard does not require admin tokens, as far as I have tested.

File: docs/checks.md
Comment (pnacht) [inline docs/checks.md:1]: Two suggestions:
- replace all cases of `For administrators:` with `With admin token:`. To my eye, it's more comprehensible and uses the same terminology used in the paragraph starting at L68.
- Explain how scores are handled without an admin token. That is, If a project satisfies all criteria except for "Dismiss stale reviews" but the check is run without an admin token, does the project get a 10/10?

File: docs/faq.md
Comment (pnacht) [inline docs/faq.md:95]: Doesn't Branch-Protection also run for release branches? (Release branches are detected only if the project uses GitHub Releases)

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:11]: It was a mistake, sorry. Will fix.

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md:117]: True, thanks! will fix.

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md:1]: Since Scorecard supports not only GitHub but also GitLab, I believe the checks description should be more generic. For GitHub it requires an admin token but for GitLab it might be another way of ensuring this rule for administrators.

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:95]: I am not sure, so will change the text to "branches" to be more generic.

File: docs/faq.md
Comment (pnacht) [reply docs/faq.md]: I think it can be simplified to

`[Branch-Protection page with the following options selected: "Require a pull request before merging", "Require approvals" with 2 approvers, "Require review from code owners", "Require status checks to pass before merging" with "Require branches to be up to date before merging" and at least one CI/CD check chosen. All other options are unchecked]`

File: docs/checks.md
Comment (pnacht) [reply docs/checks.md:1]: > Since Scorecard supports not only GitHub but also GitLab, I believe the checks description should be more generic. For GitHub it requires an admin token but for GitLab it might be another way of ensuring this rule for administrators.

Good point. Let's wait for input from others who may know better.

> About the second point, I understand it's already explained in this paragraph "This test has tiered scoring. Each tier must be fully satisfied to achieve points at the next tier. For example, if you fulfill the Tier 3 checks but do not fulfill all the Tier 2 checks, you will not receive any points for Tier 3.".

But does that also apply to running without a PAT? Does that mean that it's impossible to get 10/10 in Branch Protection without an admin token?

Comment (gabibguti) [reply docs/checks.md:1]: > But does that also apply to running without a PAT? Does that mean that it's impossible to get 10/10 in Branch Protection without an admin token?

No, you do not consider the settings enforced to admins when using a non-admin token. In other words, you can reach a 10/10 using a non-admin token.
_Just to make sure we are aligned, Scorecard cannot run without a Personal Access Token (PAT). Scorecard can run Branch-Protection with a PAT that does not have admin access to a repo, because you do not own the repo for example. Or, Scorecard can run Branch-Protection with a PAT that does have admin access._

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:89]: [ON HOLD] Analysing the best approach for the images...

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md]: [ON HOLD] Analysing the best approach for the images...

File: docs/checks.md
Comment (pnacht) [reply docs/checks.md:1]: > No, you do not consider the settings enforced to admins when using a non-admin token. In other words, you can reach a 10/10 using a non-admin token.

Understood. But then my point stands that this needs to be made clear in the documentation. As written, it seems a reasonable conclusion that you can't get a 10/10 without the admin token.

I see two solutions. Either add a

```markdown
- Tier 4
  - For non-administrators: No additional requirements
```

or `If running Scorecards without an admin token, all "For administrator" checks can be safely ignored, and scores will be determined as if all such requirements have been met.`

> _Just to make sure we are aligned, Scorecard cannot run without a Personal Access Token (PAT). Scorecard can run Branch-Protection with a PAT that does not have admin access to a repo, because you do not own the repo for example. Or, Scorecard can run Branch-Protection with a PAT that does have admin access._

Yes, sorry, I misspoke. Was thinking in terms of the Action, which doesn't require a PAT because it has the GITHUB_TOKEN. But yeah, I meant "admin token".

File: docs/faq.md
Comment (diogoteles08) [inline docs/faq.md:87]: I think these paragraph with the images would fit better *after* the "It's important to reiterate [...]" paragraph and the table, in a form of example/conclusion.

File: docs/checks.md
Comment (diogoteles08) [reply docs/checks.md:1]: I agree that this case should be clearer. At the [original documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection) they have already made it clear, actually. But I would also add something similar in this FAQ answer.
And I think that Pedro's explanation (`If running Scorecards without [...]`) is even clearer than the one currently on the documentation.

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md]: @pnacht Please review.

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:87]: I rather leave the current structure because the important question I wanna answer is how to setup the branch protection settings, so should be the first information possible. If the user is not getting a perfect score, then they would look at the tier based score table to understand why.

File: docs/faq.md
Comment (pnacht) [reply docs/faq.md]: Looks good.

File: docs/faq.md
Comment (diogoteles08) [reply docs/faq.md:87]: Hmm I see your point. Yep, makes sense

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md:1]: The documentation already states that `"If Scorecard is run without an administrative access token, the requirements that specify “For administrators” are ignored."`. I will merge this last sentence `and scores will be determined as if all such requirements have been met.` to explicitly mention that it does not affect the score. I will not reiterate this in the FAQ question though as it is already stated in the check description.

File: docs/faq.md
Comment (diogoteles08) [reply docs/faq.md]: LGTM

File: docs/checks.md
Comment (diogoteles08) [reply docs/checks.md:1]: Yep, I'm ok with this idea. The PR was not yet updated with this change tho, just fwi

Comment (pnacht) [reply docs/checks.md:1]: LGTM

Comment (gabibguti) [reply docs/checks.md:1]: True, just pushed the commit

File: docs/checks.md
Comment (diogoteles08) [inline docs/checks.md:112]: I think this actually gives 8/10.
I tested it using Scorecard 4.10.4 on project flutter/packages

**Review by diogoteles08 — CHANGES_REQUESTED** [review]

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md:112]: You're right. As per my tests, the points by Tier are a bit weird. In my understanding, it should be what I wrote in the documentation, but for some reason it's not.

Comment (gabibguti) [reply docs/checks.md:112]: Documenting here the problems with the score:
- Using non-admin token, when you reach tier 3 you have 8/10 and should be 7/10.
- Using non-admin token, when you reach tier 3 you have 8/10. When enabling "Require review from code owners" you still get an 8/10.
- Using admin token, when you reach tier 1 you have 4/10 and should be 3/10.
- Using admin token, when you reach tier 2 and enable "Require at least 1 reviewer for approval before merging " you have 5/10. When enabling "For administrators: Require branch to be up to date before merging" you still get a 5/10.
- Using admin token, when you reach tier 3 you have 8/10 and should be 7/10.
- Using admin token, when you reach tier 3 you have 8/10. When enabling "Require at least 2 reviewers for approval before merging" you still get an 8/10.
- Using admin token, when you reach tier 3 you have 8/10. When enabling "Require review from code owners" you still get an 8/10.

Comment (gabibguti) [reply docs/checks.md:112]: What I can change right now so the documentation makes more sense is:
- Tier 1 (3/10 or 4/10 points)
- Tier 3 (8/10 points)
But really this scoring should be fixed.

Comment (pnacht) [reply docs/checks.md:112]: I think some clarity from @laurentsimon or other maintainers could be useful here.

Comment (gabibguti) [reply docs/checks.md:112]: > For me, the score should be count 1 point for each setting but when using non-admin token if you reached all the tier settings you could do you get the max points for that tier.

@laurentsimon Do you agree with this scoring?

Comment (laurentsimon) [reply docs/checks.md:112]: > Using non-admin token, when you reach tier 3 you have 8/10 and should be 7/10.

That's because by adding enabling more setting in the tier, it does not update?

I think what happened is that we defined some levels, but over time more settings were introduced, so enabling certain settings does not always have the intended effect on the score.. because it's not granular enough.

I'n not against fixing the scores. Sometimes I wonder if we should have score from 1-100 :)

Comment (gabibguti) [reply docs/checks.md:112]: > That's because by adding enabling more setting in the tier, it does not update the score?

That's because the user enables 3 settings for tier 1, 3 settings for tier 2, and 1 setting for tier 3. I believe the user should get 7/10 at this point, but for some reason gets 8/10. So, it does update the score, but 1 more point than necessary.

Comment (gabibguti) [reply docs/checks.md:112]: https://github.com/ossf/scorecard/issues/2796

Comment (gabibguti) [reply docs/checks.md:112]: We can further discuss the score problem in the new issue. For the documentation to make sense for now I will add the following changes:
> What I can change right now so the documentation makes more sense is:
> - Tier 1 (3/10 or 4/10 points)
> - Tier 3 (8/10 points)

Comment (gabibguti) [reply docs/checks.md]: Please reopen if you disagree

**Review by diogoteles08 — APPROVED** [review]

File: docs/checks.md
Comment (pnacht) [reply docs/checks.md]: I'd suggest "(3/10 with admin token, 4/10 without)" or something.

**Comment by gabibguti** [comment]
Hey @laurentsimon, would you be able to review this PR?

File: docs/faq.md
Comment (spencerschrock) [inline docs/faq.md:91]: Can we pick a different example status check for these pictures?
I don't want to imply that someone needs to make `Scorecard analysis` as one of the required status checks before merging.

Even just changing it with `Inspect Elements` for the purpose of the screenshot is fine

File: docs/faq.md
Comment (spencerschrock) [inline docs/faq.md:108]: these tiers will need to be changed to reflect the actual tiers (see comments in checks.md)

**Review by spencerschrock — COMMENTED** [review]
Please make your changes in `docs/checks/internal/checks.yaml` and not in `docs/checks.md` directly.

**Comment by spencerschrock** [comment]
It may also be worth documenting how to make a token to check the admin settings, as I couldn't figure it out in the few minutes I spent experimenting (although perhaps some org settings are blocking me). I'll need to check with @laurentsimon

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md:68]: Ah yes! Seems like `RequiresStrictStatusChecks` is missing indeed. I think we should add both `RequiresStatusChecks` and `RequiresStrictStatusChecks` to the list tho. And use a slightly different naming in the list: `RequiresStatusChecks` and `UpToDateBeforeMerge`.

---

**Why we should add both `RequiresStatusChecks` and `RequiresStrictStatusChecks` to the list**

As I understand from [GitHub's Branch Protection Inputs doc](https://docs.github.com/en/graphql/reference/input-objects#createbranchprotectionruleinput), `requiresStatusChecks` and `requiresStrictStatusChecks` are different and both rules are fetched and used for scoring when using an admin token.

- `requiresStatusChecks`: Are status checks required to update matching branches.
- `requiresStrictStatusChecks `: Are branches required to be up to date before merging.

---

**Why use the different naming `RequiresStatusChecks` and `UpToDateBeforeMerge`**

In `branch.go` Scorecard creates it's own branch rules data structure [`BranchProtectionRule`](https://github.com/ossf/scorecard/blob/main/clients/branch.go#L25).

In `branches.go` Scorecard maps GitHub's branch rules stored in [`branchProtectionRule`](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L80) to it's own structure at [`branch.BranchProtectionRule`](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L99).

That's when:
- [`isAdminEnforced` from GitHub becomes `EnforceAdmins` in Scorecard](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L188).
- [`dismissesStaleReviews` from GitHub becomes `DismissStaleReviews` in Scorecard](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L190).
- [`requireLastPushApproval` from GitHub becomes `RequireLastPushApproval` in Scorecard.](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L189)
- [`requiresStatusChecks` from GitHub becomes `RequiresStatusChecks` in Scorecard.](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L192)
- [`requiresStrictStatusChecks` from GitHub becomes `UpToDateBeforeMerge` in Scorecard.](https://github.com/ossf/scorecard/blob/main/clients/githubrepo/branches.go#L193)

I believe in this list we are using the Scorecard naming, that's why all names are slightly different then GitHub's inputs names. And I really missed `UpToDateBeforeMerge` in the list.

Comment (gabibguti) [reply docs/checks.md:68]: > RequiresStatusChecks is admin protected but RequiredStatusCheckContexts is not. So I'm not sure we need this in the list of admin setings.

Accoording to [GitHub's Branch Protection Inputs doc](https://docs.github.com/en/graphql/reference/input-objects#createbranchprotectionruleinput), the rules are different.

- requiredStatusCheckContexts (non-admin): List of required status check contexts that must pass for commits to be accepted to matching branches.
- requiresStatusChecks (admin-only): Are status checks required to update matching branches.

One rule lists the status checks and the other rule see if they are required or not. So, we can view the list the status checks with non-admin token, but we wanna ensure they are actually required for merging, which we do when using admin token.

File: docs/checks.md
Comment (spencerschrock) [reply docs/checks.md:102]: This is a bug in the GitHub graphQL API, where  `requiresStrictStatusChecks` returns `true` if `requiresStatusChecks` is `false`. But in reality there's no way to have strict without the normal one.

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md]: **Explaining why I think "dismiss stale reviews" is scored in Tier 5:**

In `adminThoroughReviewProtection` method, we verify `DismissStaleReviews` to count it or not

https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L406

This score is saved in `adminThroughReview`

https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L80

In `computeAdminThoroughReviewScore`, we count the `adminThroughReview` score in the final Branch-Protection score  https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L153 

But we only count it on Tier 5  https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L241

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md]: **(Similar explanation as before) Explaining why I think "codeowners" is scored in Tier 4:**

In `codeownerBranchProtection` method, we verify `RequireCodeOwnerReviews` to count it or not

[https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L452](https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L406)

This score is saved in `codeownerReview`

[https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L81](https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L80)

In `computeCodeownerThoroughReviewScore`, we count the `codeownerReview` score in the final Branch-Protection score  [https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L171](https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L153)

But we only count it on Tier 4  [https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L230](https://github.com/ossf/scorecard/blob/main/checks/evaluation/branch_protection.go#L241)

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:91]: Sure!

**Comment by gabibguti** [comment]
> It may also be worth documenting how to make a token to check the admin settings, as I couldn't figure it out in the few minutes I spent experimenting (although perhaps some org settings are blocking me). I'll need to check with @laurentsimon

True, I agree would be a good documentation.

File: docs/faq.md
Comment (gabibguti) [reply docs/faq.md:108]: Check explanations above.

File: docs/checks.md
Comment (gabibguti) [reply docs/checks.md]: Should we leave the documentation as 3/10 score for Tier 1 and file the bug instead of changing the documentation?

**Comment by gabibguti** [comment]
The current documentation on Authorization covers how to create the token:
https://github.com/ossf/scorecard/blob/87b1c4cd6e3b1879403ee26dfbf1c8e7f3d25c02/README.md?plain=1#L231

This setting is enough to run branch protection settings as admin when it's for your own repositories. For organizations, I'm not sure. Since Scorecard docs main target seems to be usage for individual repos, I think we're good.

File: docs/checks.md
Comment (spencerschrock) [reply docs/checks.md]: Sorry, my original comment was mistaken

File: docs/checks.md
Comment (spencerschrock) [reply docs/checks.md]: In my opinion yes.

Comment (gabibguti) [reply docs/checks.md]: https://github.com/ossf/scorecard/issues/3250

**Review by spencerschrock — APPROVED** [review]
Thanks

**Pull request merged**
